src/libostree/ostree-fetcher.c

```diff
@@ -186,11 +196,34 @@ pending_task_compare (gconstpointer a,
          (priority_a < priority_b) ? -1 : 1;
 }
 
+static gboolean
+idle_process_pending_queue (gpointer user_data)
+{
+  ThreadClosure *thread_closure = user_data;
+
+  thread_closure->idle_process_queue_source = NULL;
+  session_thread_process_pending_queue (thread_closure);
+  return FALSE;
+}
+
 static void
 pending_uri_free (OstreeFetcherPendingURI *pending)
 {
-  g_hash_table_remove (pending->thread_closure->outstanding, pending);
+  ThreadClosure *thread_closure = pending->thread_closure;
+
+  /* Now we need to reprocess the queue, since this may have cleared
+   * it up.  We do it asynchronously to avoid possible reentrancy.
+   */
+  if (thread_closure->idle_process_queue_source == NULL)
+    {
+      GSource *timeout_src = g_timeout_source_new (0);
+      g_source_set_callback (timeout_src, idle_process_pending_queue, thread_closure, NULL);
```

Comment:
I think I'd throw a ref on the `thread_closure` for the duration of the timeout source:
```
g_source_set_callback (timeout_src, idle_process_pending_queue,
                       thread_closure_ref (thread_clousre),
                       (GDestroyNotify) thread_closure_unref)
```